internal/driver/driver.go

```diff
@@ -467,16 +381,79 @@ func (d *Driver) fetchCommandResponse(cmdUuid string) (string, bool) {
 }
 
 func (d *Driver) AddDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is added", deviceName))
+	d.Logger.Debugf("Device %s is added", deviceName)
 	return nil
 }
 
 func (d *Driver) UpdateDevice(deviceName string, protocols map[string]models.ProtocolProperties, adminState models.AdminState) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is updated", deviceName))
+	d.Logger.Debugf("Device %s is updated", deviceName)
 	return nil
 }
 
 func (d *Driver) RemoveDevice(deviceName string, protocols map[string]models.ProtocolProperties) error {
-	d.Logger.Debug(fmt.Sprintf("Device %s is removed", deviceName))
+	d.Logger.Debugf("Device %s is removed", deviceName)
 	return nil
 }
+
+func createMqttClient(serviceConfig *ServiceConfig) (mqtt.Client, errors.EdgeX) {
+	var scheme = serviceConfig.MQTTBrokerInfo.Schema
+	var brokerUrl = serviceConfig.MQTTBrokerInfo.Host
+	var brokerPort = serviceConfig.MQTTBrokerInfo.Port
+	var authMode = serviceConfig.MQTTBrokerInfo.AuthMode
+	var secretPath = serviceConfig.MQTTBrokerInfo.CredentialsPath
+	var mqttClientId = serviceConfig.MQTTBrokerInfo.ClientId
+	var keepAlive = serviceConfig.MQTTBrokerInfo.KeepAlive
+
+	uri := &url.URL{
+		Scheme: strings.ToLower(scheme),
+		Host:   fmt.Sprintf("%s:%d", brokerUrl, brokerPort),
+	}
+
+	err := SetCredentials(uri, "init", authMode, secretPath)
+	if err != nil {
+		return nil, errors.NewCommonEdgeXWrapper(err)
+	}
+
+	var client mqtt.Client
+	for i := 1; i <= serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry; i++ {
+		client, err = mqttClient(mqttClientId, uri, keepAlive)
+		if err != nil && i == serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry {
```

Comment:
use `>=` to prevent someone assign negative int to ConnEstablishingRetry
`i == serviceConfig.MQTTBrokerInfo.ConnEstablishingRetry`

Reply:
Modified.